File: StriverA2Z/C++/2_SortingTechniques/bubble_sort.cpp
// O(n^2) - worst, avg
// O(n) - best

#include<bits/stdc++.h>

using namespace std;

void bubbleSort(int arr[], int length) {
    for (int i = length - 1; i >= 1; i--) {
        int swap = 0;
        for (int j = 0; j <= i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                int temp = arr[j + 1];
                arr[j + 1] = arr[j];
                arr[j] = temp;
                swap = 1;
            }
        }
        if (swap == 0) break;
    }
}

int main() {
    int n;
    cout << "Enter the size of array:";
    cin >> n;

    int arr[n];
    for (int i = 0; i < n; i++)cin >> arr[i];

    bubbleSort(arr, n);

    for (int x: arr) cout << x << " ";
    return 0;
}

File: StriverA2Z/Java/SortingAlgorithms/BubbleSort.java
public class BubbleSort {
    public static void main(String[] args) {
        int[] arr = {13, 46, 24, 52, 20, 9};
        bubbleSort(arr);
    }

    static void bubbleSort(int[] arr) {
        for (int i = arr.length - 1; i >= 1; i--) {
            boolean didSwap = false;
            for (int j = 0; j <= i - 1; j++) {
                if (arr[j + 1] < arr[j]) {
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                    didSwap = true;
                }
            }
            if (!didSwap) {
                return;
            }
        }

        // for printing
        for (int x : arr) {
            System.out.print(x + " ");
        }
    }
}
